Framework/Algorithms/src/CalcCountRate.cpp

```diff
@@ -0,0 +1,653 @@
+        g_log.warning()
+            << "Normalization log " << m_pNormalizationLog->name()
+            << " time lies outside of the the whole experiment time. "
+               "Log normalization impossible.\n";
+        m_pNormalizationLog = nullptr;
+        m_normalizeResult = false;
+        useLogAccuracy = false;
+      } else {
+        if (!m_tmpLogHolder) {
+          m_tmpLogHolder =
+              Kernel::make_unique<Kernel::TimeSeriesProperty<double>>(
+                  *m_pNormalizationLog->clone());
+        }
+        m_tmpLogHolder->filterByTime(runTMin, runTMax);
+        m_pNormalizationLog = m_tmpLogHolder.get();
+        m_numLogSteps = m_pNormalizationLog->realSize();
+      }
+    } else {
+      if (tLogMin > runTMin || tLogMax < runTMax) {
+        g_log.warning() << "Normalization log " << m_pNormalizationLog->name()
+                        << " time does not cover the whole experiment time. "
+                           "Log normalization impossible.\n";
+        m_pNormalizationLog = nullptr;
+        m_normalizeResult = false;
+        useLogAccuracy = false;
+      }
+    }
+  }
+
+  if (useLogAccuracy) {
+    m_numLogSteps = m_pNormalizationLog->realSize();
+  } else {
+    m_numLogSteps = getProperty("NumTimeSteps");
+  }
+
+  m_TRangeMin = runTMin;
+  // histogramming excludes rightmost events. Modify max limit to keep them
+  double t_epsilon = double(runTMax.totalNanoseconds()) *
+                     (1 + std::numeric_limits<double>::epsilon());
+  int64_t eps_increment =
+      static_cast<int64_t>(t_epsilon - double(runTMax.totalNanoseconds()));
+  m_TRangeMax =
+      runTMax + eps_increment; // *(1+std::numeric_limits<double>::epsilon());
+}
+
+/* Retrieve and define data search ranges from input workspace parameters and
+ * algorithm properties
+ *
+ *@param InputWorkspace -- event workspace to process. Also retrieves algorithm
+ *                         properties, relevant to the workspace.
+ *@return -- the input workspace cropped according to XMin-XMax ranges in units,
+ *           requested by the user
+ *
+*/
+void CalcCountRate::setSourceWSandXRanges(
+    DataObjects::EventWorkspace_sptr &InputWorkspace) {
+
+  std::string RangeUnits = getProperty("RangeUnits");
+  auto axis = InputWorkspace->getAxis(0);
+  const auto unit = axis->unit();
+
+  API::MatrixWorkspace_sptr wst;
+  if (unit->unitID() != RangeUnits) {
+    auto conv = createChildAlgorithm("ConvertUnits", 0, 1);
+    std::string wsName = InputWorkspace->name();
+    if (wsName.empty()) {
+      wsName = "_CountRate_UnitsConverted";
+    } else {
+      wsName = "_" + wsName + "_converted";
+    }
+
+    conv->setProperty("InputWorkspace", InputWorkspace);
+    conv->setPropertyValue("OutputWorkspace", wsName);
+    std::string Emode = getProperty("Emode");
+    conv->setProperty("Emode", Emode);
+    conv->setProperty("Target", RangeUnits);
+
+    conv->execute();
+    wst = conv->getProperty("OutputWorkspace");
```

Comment:
I expect it to throw otherwise , though always unsure on when it does throw and when it does not. will be no harm to add the check.